File: packages/sdk-android/src/main/kotlin/com/mobiledevmcp/adapters/LogAdapter.kt
/**
 * LogAdapter - Handles log capturing for Android
 */

package com.mobiledevmcp.adapters

import java.text.SimpleDateFormat
import java.util.*

data class LogEntry(
    val level: String,
    val message: String,
    val timestamp: Date,
    val tag: String? = null
)

data class ErrorEntry(
    val message: String,
    val stack: String?,
    val timestamp: Date
)

class LogAdapter {
    
    private val logs = mutableListOf<LogEntry>()
    private val errors = mutableListOf<ErrorEntry>()
    private val maxLogs = 500
    private val maxErrors = 50
    private var isEnabled = false
    
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }
    
    fun enable() {
        isEnabled = true
        setupUncaughtExceptionHandler()
    }
    
    fun addLog(level: String, message: String, tag: String? = null) {
        if (!isEnabled) return
        
        synchronized(logs) {
            logs.add(LogEntry(level, message, Date(), tag))
            if (logs.size > maxLogs) {
                logs.removeAt(0)
            }
        }
    }
    
    fun addError(throwable: Throwable) {
        synchronized(errors) {
            errors.add(ErrorEntry(
                message = throwable.message ?: throwable.javaClass.simpleName,
                stack = throwable.stackTraceToString(),
                timestamp = Date()
            ))
            if (errors.size > maxErrors) {
                errors.removeAt(0)
            }
        }
    }
    
    fun getLogs(params: Map<String, Any?>): Map<String, Any> {
        val limit = (params["limit"] as? Number)?.toInt() ?: 100
        val level = params["level"] as? String
        val filter = params["filter"] as? String
        
        var filtered = logs.toList()
        
        level?.let { minLevel ->
            val levels = listOf("debug", "info", "warn", "error")
            val minIndex = levels.indexOf(minLevel)
            if (minIndex >= 0) {
                filtered = filtered.filter { 
                    levels.indexOf(it.level) >= minIndex 
                }
            }
        }
        
        filter?.let { pattern ->
            val regex = Regex(pattern, RegexOption.IGNORE_CASE)
            filtered = filtered.filter { regex.containsMatchIn(it.message) }
        }
        
        val limitedResults = filtered.takeLast(limit)
        
        return mapOf(
            "logs" to limitedResults.map { log ->
                mapOf(
                    "level" to log.level,
                    "message" to log.message,
                    "tag" to log.tag,
                    "timestamp" to dateFormat.format(log.timestamp)
                )
            },
            "total" to logs.size,
            "filtered" to limitedResults.size
        )
    }
    
    fun getRecentErrors(params: Map<String, Any?>): Map<String, Any> {
        val limit = (params["limit"] as? Number)?.toInt() ?: 10
        
        val limitedResults = errors.takeLast(limit)
        
        return mapOf(
            "errors" to limitedResults.map { error ->
                mapOf(
                    "message" to error.message,
                    "stack" to error.stack,
                    "timestamp" to dateFormat.format(error.timestamp)
                )
            },
            "total" to errors.size
        )
    }
    
    private fun setupUncaughtExceptionHandler() {
        val defaultHandler = Thread.getDefaultUncaughtExceptionHandler()
        
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            addError(throwable)
            defaultHandler?.uncaughtException(thread, throwable)
        }
    }
}

File: packages/sdk-ios/Sources/MobileDevMCP/Adapters/LogAdapter.swift
/**
 * LogAdapter - Handles log capturing for iOS
 */

import Foundation

struct LogEntry {
    let level: String
    let message: String
    let timestamp: Date
}

struct ErrorEntry {
    let message: String
    let stack: String?
    let timestamp: Date
}

class LogAdapter {
    
    private var logs: [LogEntry] = []
    private var errors: [ErrorEntry] = []
    private let maxLogs = 500
    private let maxErrors = 50
    private var isEnabled = false
    
    func enable() {
        isEnabled = true
        // Note: Intercepting print() requires custom implementation
        // This provides basic log storage
        print("[MCP SDK] Log capture enabled")
    }
    
    func addLog(level: String, message: String) {
        guard isEnabled else { return }
        
        let entry = LogEntry(
            level: level,
            message: message,
            timestamp: Date()
        )
        
        logs.append(entry)
        if logs.count > maxLogs {
            logs = Array(logs.suffix(maxLogs))
        }
    }
    
    func addError(_ error: Error) {
        let entry = ErrorEntry(
            message: error.localizedDescription,
            stack: Thread.callStackSymbols.joined(separator: "\n"),
            timestamp: Date()
        )
        
        errors.append(entry)
        if errors.count > maxErrors {
            errors = Array(errors.suffix(maxErrors))
        }
    }
    
    func getLogs(params: [String: Any]) -> [String: Any] {
        let limit = params["limit"] as? Int ?? 100
        let level = params["level"] as? String
        
        var filtered = logs
        
        if let level = level {
            let levels = ["debug", "info", "warn", "error"]
            if let minIndex = levels.firstIndex(of: level) {
                filtered = filtered.filter { entry in
                    guard let entryIndex = levels.firstIndex(of: entry.level) else { return true }
                    return entryIndex >= minIndex
                }
            }
        }
        
        return [
            "logs": filtered.suffix(limit).map { log in
                [
                    "level": log.level,
                    "message": log.message,
                    "timestamp": ISO8601DateFormatter().string(from: log.timestamp)
                ]
            },
            "total": logs.count,
            "filtered": filtered.count
        ]
    }
    
    func getRecentErrors(params: [String: Any]) -> [String: Any] {
        let limit = params["limit"] as? Int ?? 10
        
        return [
            "errors": errors.suffix(limit).map { error in
                [
                    "message": error.message,
                    "stack": error.stack ?? "",
                    "timestamp": ISO8601DateFormatter().string(from: error.timestamp)
                ]
            },
            "total": errors.count
        ]
    }
}
